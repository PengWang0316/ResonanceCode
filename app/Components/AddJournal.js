import React, { Component } from "react";
import Util from "../apis/Util";
import Loading from "./Loading"
import JournalContent from "./JournalContent";
import ReadingSearchAndList from "./ReadingSearchAndList";
import { isLogin } from "../apis/LoginApi";
import DatabaseApi from "../apis/DatabaseApi";
import PropTypes from "prop-types";

class AddJournal extends Component {

  componentWillMount(){
    // Initial variables
    this.readings = this.props.journalData ? this.props.journalData.readingIds : {}; // keep which readings should be attached on. Format is like {readingId: null}
    this.contents = {}; // Keep content keys and content
    this.contentKeyIndex = 0; // Generate keys for different contents
    this.contentIndexs = {}; // Use to track the index of content component in the array. Delete function needs it. The format is {contentKey: index}
    this.jounalId = this.props.journalData ? this.props.journalData._id : null; // Keeping journal id for update
    this.userId = isLogin(document).userid;

    this.state={
      isWriting: false,
      journalDate: this.props.journalData ? Util.getDateString(this.props.journalData.date) : Util.getCurrentDateString(),
      isDateCorrect: true,
      isEmptyReading: this.props.journalData ? false : true,
      contentComponentArray: [], // keep content component
      addJournalContent: "overview",
      pingPongState: this.props.journalData ? this.props.journalData.ping_pong_state : "Neutral"
    };

    console.log("Journal before alter: ", this.props.journalData);
    // If journal data exsit, get content
    if(this.props.journalData){
      // this.oldContentKeys = []; // keep original content keys
      this.oldReadingIds = Object.keys(this.readings); // keep original reading ids

      // delete unnecessary properties from journal object
      let journal = this.props.journalData;
      this.journalUserId = journal.user_id;
      delete journal.date;
      delete journal.user_id;
      delete journal.ping_pong_state;
      delete journal._id;
      delete journal.readingIds;
      let keyRegExp = /-(\d+)$/; // The regular expression for subtract suffixs
      console.log("Journal after alter: ", this.props.journalData);
      Object.keys(journal).map((key)=>{
        // this.oldContentKeys.push(key); // keeping the exsit keys in an array for delete function
        // handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared)
        // Also have to save the largest key number and use it to countine
        // journalData property is like {'overview-0': 'test overview', 'overview-0-isShared': true}
        // if key is isShared do not do anything
        let matchResult = key.match(keyRegExp);
        if (matchResult){
          if (this.contentKeyIndex < matchResult[1]) this.contentKeyIndex = matchResult[1]; // keep the max index number and user can contiune add new content
          this.handleAddContentClick(journal[key], key.replace(keyRegExp, ""), key, matchResult[1], journal[`${key}-isShared`]);
          // console.log("isShared: ",journal[`${key}-isShared`]);
        }
      });
      this.contentKeyIndex++;
    }

  }

  handleChange(event, element){
    let newState={};
    newState[element]=event.target.value;
    if(element=="journalDate") newState.isDateCorrect = Util.matchDateFormat(newState[element]);
    this.setState(newState);
  }

  handleChangeCallback(contentKey, contentText){
    console.log(contentKey, contentText);
    // update content in state
    this.contents[contentKey]=contentText;
  }

  handleDeleteContentCallback(contentKey){
    // Remove content from state
    if(this.contents.hasOwnProperty(contentKey)){
      delete this.contents[contentKey];
      delete this.contents[`${contentKey}-isShared`];
    }
    // Remove content component
    delete this.contentIndexs[contentKey];
    this.setComponentToStateArray();
    // this.state.contentComponentArray.splice(this.contentIndexs[contentKey],1);
    // this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSharedBoxChangeCallback(contentKey, isShared){
    this.contents[`${contentKey}-isShared`] = isShared;
  }

  handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared){
    // put component in array in order to show
    newContentKey = newContentKey ? newContentKey : `${this.state.addJournalContent}-${this.contentKeyIndex}`;
    // put the component in the track object
    this.contentIndexs[newContentKey]=<JournalContent key={contentKeyIndex ? contentKeyIndex : this.contentKeyIndex++} newContent={addJournalContent ? addJournalContent : ""} newContentName={newContentName ? newContentName : this.state.addJournalContent} newContentKey={newContentKey} handleChangeCallback={(contentKey, contentText)=>{this.handleChangeCallback(contentKey, contentText);}} handleDeleteContentCallback={(contentKey)=>{this.handleDeleteContentCallback(contentKey);}} handleSharedBoxChangeCallback={(contentKey, isShared)=>{this.handleSharedBoxChangeCallback(contentKey, isShared)}} isShared={isShared ? true : false} />;
    this.setComponentToStateArray();
    // this.setState({contentComponentArray: this.state.contentComponentArray});

    // also have to put content title to state in order to track content and update
    this.contents[newContentKey] = addJournalContent ? addJournalContent : "";
    this.contents[`${newContentKey}-isShared`] = isShared ? true : false;
  }

  setComponentToStateArray(){
    // transfor object to array in order to display
    this.state.contentComponentArray=[];
    Object.keys(this.contentIndexs).map((element)=>{
      this.state.contentComponentArray.push(this.contentIndexs[element]);
    });
    this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSubmit(event, isUpdate){
    event.preventDefault();
    // console.log("submit: contents:",this.contents);
    console.log("journal date:",this.state.journalDate);
    // Assemble a journal object for save
    let readingIdArray = [];
    Object.keys(this.readings).map((key)=>{readingIdArray.push(key)});
    let journal = Object.assign({_id:this.jounalId, date: new Date(this.state.journalDate), ping_pong_state: this.state.pingPongState, user_id: this.userId, readingIds: readingIdArray}, this.contents);

    if(isUpdate){
      console.log("submit: journal:", journal);
      // console.log("oldContentKeys:", this.oldContentKeys);
      console.log("oldReadingKeys:", this.oldReadingIds);
      // Assemble two arrays for deletion contents and reading
      let deleteContents = [];
      let deleteReadingIds = [];
      // this.oldContentKeys.map((element)=>{if(!this.contents.hasOwnProperty(element)) deleteContents.push(element);});
      this.oldReadingIds.map((element)=>{if(!this.readings.hasOwnProperty(element)) deleteReadingIds.push(element);});
      // journal.deleteContents = deleteContents; //todo delete deleteContents and oldContentKeys
      journal.deleteReadingIds = deleteReadingIds;
      console.log("delete contents: ",deleteContents);
      console.log("delete reading ids : ",deleteReadingIds);
      DatabaseApi.updateJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }else{
      DatabaseApi.createJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }
    /**/
    this.setState({isWriting: true});
    console.log("submit: journal:", journal);
  }

  /* For ReadingSearchAndList callback */
  handleAttachReadingCallback(readingId){
    this.readings[readingId] = null;
    this.setState({isEmptyReading: false});
  }

  handleDetachAttachReadingCallback(readingId){
    delete this.readings[readingId];
    this.setState({isEmptyReading: Object.keys(this.readings).length===0});
  }

  handleCancel(){
    this.props.history.push("/reading");
  }

  handleDelete(){
    console.log("Delete journal!");
  }

  render(){
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        <div className="titleDiv">{this.props.journalData ? "Update journal for readings" : "Add a new journal for readings"}</div>
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event, this.props.journalData);}}>

          <div className="text-right bottom-btn-div">
            {(!this.props.journalData || this.userId == this.journalUserId) &&
              <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
            }
            {(this.props.journalData && this.userId == this.journalUserId) &&
              <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
            }
            <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
          </div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">

                <input className={this.state.isDateCorrect?"form-control":"form-control form-control-warning"} type="text" placeholder="mm/dd/yyyy" id="journalDate" value={this.state.journalDate} onChange={(event)=>{this.handleChange(event, "journalDate")}} />
                {!this.state.isDateCorrect && <span className="glyphicon glyphicon-warning-sign form-control-feedback form-control-warning-span"></span>}

              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            {/* New content goes here */}
            {this.state.contentComponentArray}

            {/* Add content button and drop list */}
            <div className="row addJournalContentDiv">

              <div onClick={()=>{this.handleAddContentClick();}} className="addJournalContentBtnDiv col-sm-6"><i className="fa fa-plus-square" /> Add one content for your journal</div>
              <div className="col-sm-6">
                <select  className="form-control" onChange={(event)=>{this.handleChange(event,"addJournalContent");}}>
                  <option value="overview">Overview</option>
                  <option value="internal_environment">Internal Environment</option>
                  <option value="relational_environment">Relational Environment</option>
                  <option value="physical_environment">Physical Environment</option>
                  <option value="creative_vector">Creative Vector</option>
                  <option value="synchronicities">Synchronicities</option>
                  <option value="dreams">Dreams</option>
                  <option value="other">Other</option>
                </select>

              </div>

            </div>


            {/*  Start reading search function */}
            <ReadingSearchAndList readings={this.props.journalData ? this.readings : null} attachReadingCallback={(readingId)=>{this.handleAttachReadingCallback(readingId);}} detachReadingCallback={(readingId)=>{this.handleDetachAttachReadingCallback(readingId);}} />

              <div className="text-right bottom-btn-div">
                {(!this.props.journalData || this.userId == this.journalUserId) &&
                  <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
                }
                {(this.props.journalData && this.userId == this.journalUserId) &&
                  <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
                }
                <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
              </div>
          </form>

        </div>
    );
  }


}
AddJournal.propTypes={
  journalData: PropTypes.object,
  history: PropTypes.object
};
export default AddJournal;
