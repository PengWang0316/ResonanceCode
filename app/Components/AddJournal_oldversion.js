import React, { Component } from "react";
import QueryString from "query-string";
import PropTypes from "prop-types";
import Util from "../apis/Util";
// import api from "../apis/api";
import DatabaseApi from "../apis/DatabaseApi";
import Loading from "./Loading";
import LoginApi from "../apis/LoginApi";

class AddJournal_oldversion extends Component{

  componentWillMount(){

    if(this.props.location){
      let queryInfo = QueryString.parse(this.props.location.search);
      this.readingId = queryInfo.readingId;
      this.readingName = queryInfo.readingName;
      this.readingDate = Util.getDateString(queryInfo.readingDate);
      // console.log("AddJournal paper readingId, readingName, and readingDate:", `Id: ${this.readingId} Name: ${this.readingName} Date: ${this.readingDate}`);
    }
    this.userId=LoginApi.isLogin(document).userid;

    let journalData = this.props.journalData;
    if(journalData){
      this.journalId=journalData.journalId;
      // this.readingDate=journalData.readingDate;
      this.readingName=journalData.readingName;
    }
    // console.log("AddJournal page:",journalData);
    // setting state
    this.state={
      isWriting: false,
      journalDate: journalData?Util.getDateString(journalData.date):Util.getCurrentDateString(),
      pingPongState: journalData?journalData.pingPongState:"Neutral",
      overview: journalData?journalData.overview:"",
      internalEnvironment: journalData?journalData.internalEnvironment:"",
      relationalEnvironment: journalData?journalData.relationalEnvironment:"",
      physicalEnvironment: journalData?journalData.physicalEnvironment:"",
      creativeVector: journalData?journalData.creativeVector:"",
      synchronicities: journalData?journalData.synchronicities:"",
      dreams: journalData?journalData.dreams:"",
      journalId: journalData?journalData.journalId:"",
      userId: journalData?journalData.userId:"",
      readingId: journalData?journalData.readingId:""
    };
    // if(this.journalId) this.setState(this.state);
  }

  componentDidMount(){
    // console.log("did mount");
    if(this.journalId) this.setState(this.state);
  }

  handleChange(event, element){
    // console.log(event.target.value);
    let newState={};
    newState[element]=event.target.value;
    this.setState(newState);
    // console.log(newState);
  }

  handleSubmit(event){
    event.preventDefault();
    // console.log("**********submit************");

      this.setState({isWriting: true});
      // console.log("submit:",this.state);
      // assemble object for creating
      let journalObject = {
        user_id: this.userid,
        reading_id: this.readingId?this.readingId:this.state.readingId,
        date: new Date(this.state.journalDate),
        overview: this.state.overview,
        internal_environment: this.state.internalEnvironment,
        relational_environment: this.state.relationalEnvironment,
        physical_environment: this.state.physicalEnvironment,
        creative_vector: this.state.creativeVector,
        synchronicities: this.state.synchronicities,
        dreams: this.state.dreams,
        ping_pong_state:this.state.pingPongState
      };
      // console.log("journalId", this.journalId);
      if(this.journalId) DatabaseApi.updateJournal(journalObject, this.state.journalId).then((reault)=>{this.props.history.push("/reading")});
      else DatabaseApi.createJournal(journalObject).then((reault)=>{this.props.history.push("/reading")});

  }

  handleDelete(event){
    event.preventDefault();
    // console.log("**********delete************");
    DatabaseApi.deleteJournal(this.state.readingId, this.state.journalId).then((reault)=>{this.props.history.push("/reading")});
  }

  render(){
    let titleDiv;
    if (this.readingId) titleDiv = <div><span className="rcTitle">Add a new journal to your reading: </span>{this.readingName}  ({this.readingDate})</div>;
      else titleDiv = <div><span className="rcTitle">Update {this.readingName}</span></div>;
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        {titleDiv}
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event);}}>

          <div className="text-right bottom-btn-div">{(!this.readingId && this.state.userId==this.userId) && <button type="button" className="btn btn-danger btn-del" onClick={(event)=>{this.handleDelete(event);}}>Delete</button>}{(this.readingId || this.state.userId==this.userId) && <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.overview.length>0)}>{this.readingId?"Submit":"Update"}</button>}</div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">
                <input className="form-control" type="text" value={this.state.journalDate} id="journalDate" onChange={(event)=>{this.handleChange(event,"journalDate");}} />
              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="overview" className="col-form-label">Overview</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.overview} placeholder="Overview..." id="overview" onChange={(event)=>{this.handleChange(event,"overview");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="internalEnvironment" className="col-form-label">Internal Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.internalEnvironment} placeholder="Internal Environment..." id="internalEnvironment" onChange={(event)=>{this.handleChange(event,"internalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="relationalEnvironment" className="col-form-label">Relational Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.relationalEnvironment} placeholder="Relational Environment..." id="relationalEnvironment" onChange={(event)=>{this.handleChange(event,"relationalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="physicalEnvironment" className="col-form-label">Physical Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.physicalEnvironment} placeholder="Physical Environment..." id="physicalEnvironment" onChange={(event)=>{this.handleChange(event,"physicalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="creativeVector" className="col-form-label">Creative Vector</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.creativeVector} placeholder="Creative Vector..." id="creativeVector" onChange={(event)=>{this.handleChange(event,"creativeVector");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="synchronicities" className="col-form-label">Synchronicities</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.synchronicities} placeholder="Synchronicities..." id="synchronicities" onChange={(event)=>{this.handleChange(event,"synchronicities");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="dreams" className="col-form-label">Dreams</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.dreams} placeholder="Dreams..." id="dreams" onChange={(event)=>{this.handleChange(event,"dreams");}} />
              </div>
            </div>

          <div className="text-right bottom-btn-div">{(!this.readingId && this.state.userId==this.userId) && <button className="btn btn-danger btn-del" onClick={(event)=>{this.handleDelete(event);}}>Delete</button>}{(this.readingId || this.state.userId==this.userId) && <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.overview.length>0)}>{this.readingId?"Submit":"Update"}</button>}</div>

        </form>

      </div>
    );
  }

}
AddJournal_oldversion.propTypes={
  journalData: PropTypes.object
};
export default AddJournal_oldversion;
