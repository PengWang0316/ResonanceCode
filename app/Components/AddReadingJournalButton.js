import React from 'react';
import { Link } from 'react-router-dom';
import jQuery from 'jquery';

import styles from '../styles/AddReadingJournalButton.module.css';

const handleAddBtnClick = () => {
  const isShowAddBtns = jQuery('#addReadingBtn').css('opacity') !== '0';
  const opacity = isShowAddBtns ? '0' : '1';
  const readingBottom = isShowAddBtns ? '25px' : '85px';
  const journalBottom = isShowAddBtns ? '25px' : '145px';
  jQuery('#addReadingBtn').css({ opacity, bottom: readingBottom });
  jQuery('#addJournalBtn').css({ opacity, bottom: journalBottom });
};

// $ will use jQuery in the index.html.
const handleAddReadingButtonClick = () => $('#addReadingModal').modal('toggle');

const AddReadingJurnalButton = () => (
  <div role="button" tabIndex="-3" onClick={handleAddBtnClick}>
    {/* this $('#addReadingModal').modal('toggle') will use jQuery from index.html */}
    <div role="button" tabIndex="-4" onClick={handleAddReadingButtonClick} id="addReadingBtn" className={`${styles.addBtn} ${styles.addReadingBtn} text-center`}>Reading</div>
    <Link to={{ pathname: '/addJournal' }}><div id="addJournalBtn" className={`${styles.addBtn} ${styles.addJournalBtn} text-center`}>Journal</div></Link>
    <div className={`${styles.addBtn}`}><i className="fa fa-plus" /></div>
  </div>
);

export default AddReadingJurnalButton;
