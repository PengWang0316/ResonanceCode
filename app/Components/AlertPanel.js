import React from 'react';

import styles from '../styles/AlertPanel.module.css';

const AlertPanel = ({
  id = '', type = 'success', children = '', style = {}
}) => (
  <div id={id} className={`alert alert-${type} mt-3 ${styles.alertPanel}`} style={style} role="alert">
    {children}
  </div>
);
export default AlertPanel;
