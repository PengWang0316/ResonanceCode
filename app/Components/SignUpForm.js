import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import styles from '../styles/SignUpForm.module.css';
import { checkUserNameAvailable, registerNewUser } from '../actions/UserActions';

/** The sign up form component. */
export class SignUpForm extends Component {
  static propTypes = {
    registerNewUser: PropTypes.func.isRequired,
    checkUserNameAvailable: PropTypes.func.isRequired,
    user: PropTypes.objectOf(PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.bool,
      PropTypes.objectOf(PropTypes.bool)
    ])).isRequired
  };

  state = {
    userName: '',
    password: '',
    repeatPassword: '',
    isPasswordSame: true,
    isNameAvailable: false,
    isChecking: false,
    hasResult: false
  };

  /** Redirecting a user to the reading page after sign up succesfully.
    * Showing the warnning message if the username has already existed.
    * @param {object} nextProps is the object that contains next props information.
    * @return {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    if (nextProps.user.isChecked) this.setState({
      isChecking: false, hasResult: true, isNameAvailable: nextProps.user.isUsernameAvailable
    });
  }

  /** Checking whether the username has already exsited.
    * @param {string} username is a string that repersents the name a user wants.
    * @returns {null} No return.
  */
  checkUserName(username) {
    this.setState({ hasResult: false }); // lock the submit button when content was changed
    // Clearing the previous checkUserName function
    if (this.checkUserNameFunction) clearTimeout(this.checkUserNameFunction);
    this.checkUserNameFunction = setTimeout(() => {
      this.setState({ isChecking: true });
      this.props.checkUserNameAvailable(username);
      /* this.props.checkUserName(username, result => {
        this.setState({isChecking: false, hasResult: true, isNameAvailable: result});
      }); */
    }, 1000);
  }

  /** Setting up the states when a user changes the value in the input element.
    * @param {object} event is an object that comes from the input the user is reacting with.
    * @returns {null} No return.
  */
  handleInputChange = ({ target }) => {
    // comparing two passwords
    if (target.id === 'password') this.setState({ isPasswordSame: target.value === this.state.repeatPassword });
    else if (target.id === 'repeatPassword') this.setState({ isPasswordSame: target.value === this.state.password });

    // Checking availability for user name
    if (target.id === 'userName') this.checkUserName(target.value);

    this.setState({
      [target.id]: target.value
    });
  };

  /** Register a new user when a user submits the form.
    * @param {object} event is an object that comes from the input the user is reacting with.
    * @returns {null} No return.
  */
  handleSubmit = event => {
    event.preventDefault();
    this.props.registerNewUser({ username: this.state.userName, password: this.state.password });
  };

  /** Render method for the component.
    * @returns {jsx} Return the jsx for the component.
  */
  render() {
    return (
      <div>
        <div className={`text-center ${styles.signupTitle}`}>New User Sign Up</div>
        <form className="form-horizontal" onSubmit={this.handleSubmit}>

          <div className={this.state.userName !== '' && !this.state.isNameAvailable && this.state.hasResult ? 'form-group has-error has-feedback' : 'form-group'}>
            <label htmlFor="userName" className={`control-lable ${styles.signupLable}`}>User Name</label>{this.state.isChecking && <i className="fa fa-spinner fa-spin" />}{this.state.userName !== '' && !this.state.isNameAvailable && this.state.hasResult && <span className={`${styles.wrongNameInfo}`}>Name not available!</span>}
            <input onChange={this.handleInputChange} type="text" className="form-control" id="userName" placeholder="User Name..." value={this.state.userName} aria-describedby="inputError2Status" />
            <span className={`glyphicon glyphicon-remove form-control-feedback form-control-feedback-img ${styles.formControlFeedbackImg}`} aria-hidden="true" />
            <span id="inputError2Status" className="sr-only">(error)</span>
          </div>

          <div className="form-group">
            <label htmlFor="password" className={`${styles.signupLable}`}>Password</label>
            <input onChange={this.handleInputChange} type="password" className="form-control" id="password" placeholder="Password..." value={this.state.password} />
          </div>

          <div className={this.state.isPasswordSame ? 'form-group' : 'form-group has-error has-feedback'}>
            <label htmlFor="repeatPassword" className={`control-label ${styles.signupLable}`}>Repeat Password</label>{!this.state.isPasswordSame && <span className={`${styles.wrongNameInfo}`}>Passwords different!</span>}
            <input onChange={this.handleInputChange} type="password" className="form-control" id="repeatPassword" placeholder="Repeat Password..." value={this.state.repeatPassword} />
            <span className={`glyphicon glyphicon-remove form-control-feedback form-control-feedback-img ${styles.formControlFeedbackImg}`} aria-hidden="true" />
            <span id="inputError2Status" className="sr-only">(error)</span>
          </div>

          <div className="text-right form-group"><button disabled={this.state.userName === '' || this.state.password === '' || !this.state.isPasswordSame || !this.state.isNameAvailable || !this.state.hasResult} type="submit" className="btn btn-primary">Sign Up</button></div>
        </form>
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({ user: state.user });
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  checkUserNameAvailable: username => dispatch(checkUserNameAvailable(username)),
  registerNewUser: user => dispatch(registerNewUser(user))
});
export default connect(mapStateToProps, mapDispatchToProps)(SignUpForm);
