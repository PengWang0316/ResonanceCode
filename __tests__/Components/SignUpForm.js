import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { SignUpForm } from '../../app/Components/SignUpForm';

describe('SignUpForm test', () => {
  const defaultProps = {
    registerNewUser: jest.fn(),
    checkUserNameAvailable: jest.fn(),
    user: { isChecked: false, isUsernameAvaliable: true }
  };
  const getShallowComponent = (props = defaultProps) => shallow(<SignUpForm {...props} />);

  test('SignUpForm snapshot', () => expect(renderer.create(<SignUpForm {...defaultProps} />).toJSON()).toMatchSnapshot());

  test('inintial states', () => {
    const component = getShallowComponent();
    expect(component.state('userName')).toBe('');
    expect(component.state('password')).toBe('');
    expect(component.state('repeatPassword')).toBe('');
    expect(component.state('isPasswordSame')).toBeTruthy();
    expect(component.state('isNameAvailable')).toBeFalsy();
    expect(component.state('isChecking')).toBeFalsy();
    expect(component.state('hasResult')).toBeFalsy();
  });

  test('componentWillReceiveProps', () => {
    const component = getShallowComponent();
    component.setState({ isChecking: true, hasResult: false, isNameAvailable: false });
    component.setProps({ user: { isChecked: true, isUsernameAvailable: true } });
    expect(component.state('isChecking')).toBeFalsy();
    expect(component.state('hasResult')).toBeTruthy();
    expect(component.state('isNameAvailable')).toBeTruthy();
  });

  test('checkUserName', () => {
    jest.useFakeTimers();
    const component = getShallowComponent();
    component.setState({ isChecking: false });
    const mockCheckUserNameFunc = jest.fn();
    component.instance().checkUserNameFunction = mockCheckUserNameFunc;
    component.find('input').at(0).simulate('change', { target: { id: 'userName', value: 'newUserName' } });
    expect(clearTimeout).toHaveBeenLastCalledWith(mockCheckUserNameFunc);
    expect(defaultProps.checkUserNameAvailable).not.toHaveBeenCalled();
    jest.runAllTimers();
    expect(component.state('isChecking')).toBeTruthy();
    expect(defaultProps.checkUserNameAvailable).toHaveBeenLastCalledWith('newUserName');
  });

  test('handleInputChange password', () => {
    const component = getShallowComponent();
    component.setState({ repeatPassword: 'aaa', isPasswordSame: true });
    component.find('input').at(0).simulate('change', { target: { id: 'password', value: 'password' } });
    expect(component.state('password')).toBe('password');
    expect(component.state('isPasswordSame')).toBeFalsy();
  });

  test('handleInputChange repeatPassword', () => {
    const component = getShallowComponent();
    component.setState({ password: 'aaa', isPasswordSame: false });
    component.find('input').at(0).simulate('change', { target: { id: 'repeatPassword', value: 'aaa' } });
    expect(component.state('repeatPassword')).toBe('aaa');
    expect(component.state('isPasswordSame')).toBeTruthy();
  });

  test('handleSubmit', () => {
    const component = getShallowComponent();
    component.setState({ userName: 'username', password: 'password' });
    const mockFunc = jest.fn();
    component.find('form').simulate('submit', { preventDefault: mockFunc });
    expect(mockFunc).toHaveBeenCalledTimes(1);
    expect(defaultProps.registerNewUser).toHaveBeenLastCalledWith({ username: 'username', password: 'password' });
  });
});
