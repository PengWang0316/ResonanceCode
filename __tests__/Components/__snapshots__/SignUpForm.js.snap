// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm test SignUpForm snapshot 1`] = `
<div>
  <div
    className="text-center signupTitle"
  >
    New User Sign Up
  </div>
  <form
    className="form-horizontal"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        className="control-lable signupLable"
        htmlFor="userName"
      >
        User Name
      </label>
      <input
        aria-describedby="inputError2Status"
        className="form-control"
        id="userName"
        onChange={[Function]}
        placeholder="User Name..."
        type="text"
        value=""
      />
      <span
        aria-hidden="true"
        className="glyphicon glyphicon-remove form-control-feedback form-control-feedback-img formControlFeedbackImg"
      />
      <span
        className="sr-only"
        id="inputError2Status"
      >
        (error)
      </span>
    </div>
    <div
      className="form-group"
    >
      <label
        className="signupLable"
        htmlFor="password"
      >
        Password
      </label>
      <input
        className="form-control"
        id="password"
        onChange={[Function]}
        placeholder="Password..."
        type="password"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="control-label signupLable"
        htmlFor="repeatPassword"
      >
        Repeat Password
      </label>
      <input
        className="form-control"
        id="repeatPassword"
        onChange={[Function]}
        placeholder="Repeat Password..."
        type="password"
        value=""
      />
      <span
        aria-hidden="true"
        className="glyphicon glyphicon-remove form-control-feedback form-control-feedback-img formControlFeedbackImg"
      />
      <span
        className="sr-only"
        id="inputError2Status"
      >
        (error)
      </span>
    </div>
    <div
      className="text-right form-group"
    >
      <button
        className="btn btn-primary"
        disabled={true}
        type="submit"
      >
        Sign Up
      </button>
    </div>
  </form>
</div>
`;
